File: server/lib/download.js
import ytdl from 'youtube-dl-exec';
import yts from 'yt-search'
import fs from 'fs';
import path from 'path';
import logger from '../utils/logger.js';
import ffmpeg from 'fluent-ffmpeg';
import { getFfmpegPath, getCookiesPath } from '../utils/utils.js';
import axios from 'axios';
import cacheManager from './cacheManager.js';
import fsHelper from '../utils/helper/fs-helper.js';
import { DEFAULT_TRACKS_LOCATION } from '../utils/constant.js';
import SoundCloud from './soundcloud.js';
import { createDownloadLinks } from '../utils/crypto.js';

ffmpeg.setFfmpegPath(getFfmpegPath());
class MyDownloader {
    
    async testCookies() {
        try {
            const cookiesPath = getCookiesPath();
            logger.info('Testing YouTube cookies...');
            
            // Test with a known age-restricted video that requires authentication
            const testUrl = 'https://www.youtube.com/watch?v=tPx-7Grk_UY';
            const result = await ytdl(testUrl, {
                dumpSingleJson: true,
                cookies: cookiesPath,
                verbose: true
            });
            
            logger.info('YouTube cookies are working correctly');
            return true;
        } catch (error) {
            logger.error('YouTube cookie test failed:', error);
            return false;
        }
    }

    async getVideoDetail(name, artistName) {
        try {
            const r = await yts({ query: `${name} - ${artistName} official audio song music`, category: 'music' });
            if (r.videos?.length === 0) {
                throw new Error('No video found for the given name and artist');
            }

            return r.videos[0];
        } catch (error) {
            logger.error("Error getting details: " + error.message);
            throw error;
        }
    }

    async downloadVideo(url, title, outputPath = DEFAULT_TRACKS_LOCATION) {
        // Check if song exists in cache first
        const cachedPath = cacheManager.getFromCache(title);
        if (cachedPath) {
            logger.info(`Using cached version of: ${title}`);
            return { url: cachedPath };
        }

        // Ensure tracks directory exists
        if (!fsHelper.exists(outputPath)) {
            fsHelper.createDirectory(outputPath);
            logger.info(`Created directory: ${outputPath}`);
        }

        // Get consistent file path
        const outputFilePath = cacheManager.getOriginalPath(title);
        logger.info(`Downloading ${title} to ${outputFilePath}`);
        try {
            const cookiesPath = getCookiesPath();
            
            const cookiesContent = fs.readFileSync(cookiesPath, 'utf8');
            const cookieLines = cookiesContent.split('\n').filter(line => 
                line.trim() && !line.startsWith('#') && line.includes('.youtube.com')
            );

            if (cookieLines.length === 0) {
                logger.warn('No YouTube cookies found, download may fail. Please add cookies to cookies.txt');
            } else {
                logger.info(`Using ${cookieLines.length} YouTube cookies for download`);
            }

            const options = {
                output: outputFilePath,
                extractAudio: true,
                audioFormat: 'mp3',
                audioQuality: 6,
                noCallHome: true,
                noCheckCertificate: true,
                preferFreeFormats: true,
                ffmpegLocation: getFfmpegPath(),
                cookies: cookiesPath,
                verbose: true
            };

            logger.info(`Downloading ${title} to ${outputFilePath} with YouTube cookies`);
            await ytdl(url, options);
            logger.info(`Successfully downloaded ${title} to ${outputFilePath}`);
            return { url: outputFilePath }

        } catch (error) {
            logger.error(`Error downloading ${title}:`, error);
            throw error;
        }
    }

    async downloadFromUrl(url, title, outputPath = DEFAULT_TRACKS_LOCATION) {
        // Check if song exists in cache first
        const cachedPath = cacheManager.getFromCache(title);
        if (cachedPath) {
            logger.info(`Using cached version of: ${title}`);
            return { url: cachedPath };
        }

        // Ensure tracks directory exists
        if (!fsHelper.exists(outputPath)) {
            fsHelper.createDirectory(outputPath);
            logger.info(`Created directory: ${outputPath}`);
        }

        // Get consistent file paths
        const outputFilePath = cacheManager.getOriginalPath(title);
        const safeTitle = title.replace(/[<>:"/\\|?*]/g, '');
        const tempFile = path.join(outputPath, `temp_${safeTitle}.mp3`);
        logger.info(`Downloading ${title} to ${outputFilePath}`);

        try {
            logger.info(`Downloading ${title} from URL to ${outputFilePath}`);
            const response = await axios({
                method: 'get',
                url: url,
                responseType: 'stream'
            });

            const writer = fs.createWriteStream(tempFile);
            response.data.pipe(writer);

            await new Promise((resolve, reject) => {
                writer.on('finish', resolve);
                writer.on('error', reject);
            });

            await new Promise((resolve, reject) => {
                ffmpeg(tempFile)
                    .toFormat('mp3')
                    .audioQuality(6)
                    .on('end', () => {
                        fs.unlink(tempFile, (err) => {
                            if (err) logger.error('Error deleting temp file:', err);
                        });
                        logger.info(`Successfully downloaded ${title} to ${outputFilePath}`);
                        resolve();
                    })
                    .on('error', (err) => {
                        fs.unlink(tempFile, () => {
                            reject(err);
                        });
                    })
                    .save(outputFilePath);
            });

            return { url: outputFilePath };
        } catch (error) {
            logger.error(`Error downloading ${title}:`, error);
            if (fsHelper.exists(tempFile)) {
                fsHelper.delete(tempFile);
                logger.info(`Cleaned up temp file: ${tempFile}`);
            }
            throw error;
        }
    }

    async downloadSoundCloud(url, title){
        const download = new SoundCloud();
        const streamUrl = await download.fetchStreamUrl(url);
        return await this.downloadFromUrl(streamUrl, title)
    }

    async downloadJioSaavn(url, title){
        const streamUrl = createDownloadLinks(url)[3].url;
        return await this.downloadFromUrl(streamUrl, title)
    }
}

export default MyDownloader;

File: app/streaming/download.py
import yt_dlp as ytdl
import aiohttp
import asyncio
import subprocess
import os
from pathlib import Path
from app.core import logger
from app.core.utils import get_ffmpeg_path, get_cookies_path
from app.core.constants import DEFAULT_TRACKS_LOCATION
from app.core.fs_helper import fs_helper
from app.core.crypto import create_download_links
from app.streaming.cache_manager import cache_manager

class Downloader:
    async def download_video(self, url: str, title: str, output_path: str = DEFAULT_TRACKS_LOCATION):
        cached_path = cache_manager.get_from_cache(title)
        if cached_path:
            logger.info(f"Using cached version of: {title}")
            return {'url': cached_path}
        
        if not fs_helper.exists(output_path):
            fs_helper.create_directory(output_path)
            logger.info(f"Created directory: {output_path}")
        
        output_file_path = cache_manager.get_original_path(title)
        logger.info(f"Downloading {title} to {output_file_path}")
        
        try:
            cookies_path = get_cookies_path()
            
            use_cookies = False
            if os.path.exists(cookies_path):
                with open(cookies_path, 'r') as f:
                    cookies_content = f.read()
                    cookie_lines = [
                        line for line in cookies_content.split('\n')
                        if line.strip() and not line.startswith('#') and '.youtube.com' in line
                    ]
                    
                    if len(cookie_lines) > 0:
                        logger.info(f"Found {len(cookie_lines)} YouTube cookies, will try with cookies first")
                        use_cookies = True
                    else:
                        logger.warn('No valid YouTube cookies found in cookies.txt')
            else:
                logger.warn('No cookies.txt file found')
            
            download_methods = [
                {
                    'name': 'without cookies',
                    'options': {
                        'format': 'bestaudio/best',
                        'outtmpl': output_file_path,
                        'postprocessors': [{
                            'key': 'FFmpegExtractAudio',
                            'preferredcodec': 'mp3',
                            'preferredquality': '192',
                        }],
                        'ffmpeg_location': get_ffmpeg_path(),
                        'quiet': True,
                        'no_warnings': True,
                    }
                }
            ]
            
            if use_cookies:
                download_methods.append({
                    'name': 'with cookies',
                    'options': {
                        'format': 'bestaudio/best',
                        'outtmpl': output_file_path,
                        'postprocessors': [{
                            'key': 'FFmpegExtractAudio',
                            'preferredcodec': 'mp3',
                            'preferredquality': '192',
                        }],
                        'cookiefile': cookies_path,
                        'ffmpeg_location': get_ffmpeg_path(),
                        'quiet': True,
                        'no_warnings': True,
                    }
                })
            
            download_successful = False
            last_error = None
            
            for method in download_methods:
                try:
                    logger.info(f"Attempting download of {title} {method['name']}")
                    with ytdl.YoutubeDL(method['options']) as ydl:
                        await asyncio.to_thread(ydl.download, [url])
                    logger.info(f"Successfully downloaded {title} using {method['name']}")
                    download_successful = True
                    break
                except Exception as error:
                    logger.warn(f"Download failed {method['name']}: {str(error)}")
                    last_error = error
                    continue
            
            if not download_successful:
                raise Exception(f"All download methods failed. Last error: {str(last_error)}")
            
            return {'url': output_file_path}
        
        except Exception as error:
            logger.error(f"Error downloading {title}", error=str(error))
            raise error
    
    async def download_from_url(self, url: str, title: str, output_path: str = DEFAULT_TRACKS_LOCATION):
        cached_path = cache_manager.get_from_cache(title)
        if cached_path:
            logger.info(f"Using cached version of: {title}")
            return {'url': cached_path}
        
        if not fs_helper.exists(output_path):
            fs_helper.create_directory(output_path)
            logger.info(f"Created directory: {output_path}")
        
        output_file_path = cache_manager.get_original_path(title)
        safe_title = title.replace('<', '').replace('>', '').replace(':', '').replace('"', '').replace('/', '').replace('\\', '').replace('|', '').replace('?', '').replace('*', '')
        temp_file = os.path.join(output_path, f"temp_{safe_title}.mp3")
        logger.info(f"Downloading {title} to {output_file_path}")
        
        try:
            logger.info(f"Downloading {title} from URL to {output_file_path}")
            
            async with aiohttp.ClientSession() as session:
                async with session.get(url) as response:
                    with open(temp_file, 'wb') as f:
                        async for chunk in response.content.iter_chunked(8192):
                            f.write(chunk)
            
            ffmpeg_path = get_ffmpeg_path()
            process = await asyncio.create_subprocess_exec(
                ffmpeg_path, '-i', temp_file,
                '-acodec', 'libmp3lame',
                '-aq', '6',
                output_file_path,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE
            )
            await process.communicate()
            
            if fs_helper.exists(temp_file):
                fs_helper.delete(temp_file)
            
            logger.info(f"Successfully downloaded {title} to {output_file_path}")
            return {'url': output_file_path}
        
        except Exception as error:
            logger.error(f"Error downloading {title}", error=str(error))
            if fs_helper.exists(temp_file):
                fs_helper.delete(temp_file)
                logger.info(f"Cleaned up temp file: {temp_file}")
            raise error
    
    async def download_jiosaavn(self, url: str, title: str):
        stream_url = create_download_links(url)[3]['url']
        return await self.download_from_url(stream_url, title)
    
    async def download_soundcloud(self, url: str, title: str):
        return await self.download_video(url, title)

downloader = Downloader()
